lis2ds12 accelerometer alignment

avisconti opened the pull request:
Improvements:                                                                                
                                                                                
0. Aligned stmesc API to latest DT facilities                                                                               
1. Make use of latest stmemsc module lis2ds12_STdC APIs                                                                                
2. Add multi-instance support                                                                                
3. Move range and odr into DT                                                                                
                                                                                
Tested with x_nucleo_iks01a3 shield in SPI mode

avisconti committed:
driver/sensor: lis2ds12: make use of STdC definitions

Port the lis2ds12 sensor driver on top of the lis2ds12_StdC
HAL interface (in modules/hal/st/sensor/stmemsc/).

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
dts/bindings: lis2ds12: create a common DT binding file

Create a common properties file that will be included by all DT
bindings (as i2c and spi) handled by lis2ds12 driver.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
drivers/sensor: lis2ds12: Move range Kconfig property into dts

Converts lis2ds12 range options (2g, 4g, 8g, 16g) from Kconfig
to Device Tree.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

gmarull commented on drivers/sensor/lis2ds12/lis2ds12_spi.c:
```diff
@@ -94,68 +90,24 @@ static int lis2ds12_raw_write(struct lis2ds12_data *data, uint8_t reg_addr,
-	.update_reg = lis2ds12_spi_update_reg,
+stmdev_ctx_t lis2ds12_spi_ctx = {
+	.read_reg = (stmdev_read_ptr) lis2ds12_spi_read,
+	.write_reg = (stmdev_write_ptr) lis2ds12_spi_write,
```
nit: extra space after `)`

avisconti replied:
ack

gmarull commented on drivers/sensor/lis2ds12/lis2ds12.c:
```diff
@@ -267,32 +240,121 @@ static int lis2ds12_init(const struct device *dev)
+			   (void *)&lis2ds12_config_##inst.stmemsc_cfg,	\
+		},							\
+		.stmemsc_cfg.spi = {					\
+			.bus = DEVICE_DT_GET(DT_INST_BUS(inst)),	\
+			.spi_cfg = SPI_CONFIG_DT_INST(inst,		\
```
`SPI_DT_SPEC_INST_GET` should be used

avisconti replied:
It seems fine, as it is a code semplification. But I have to change it into stmemsc_cfg_spi structure, which is common to many ST sensor now. So, I have to do it as a preliminary commit for those sensors first, and I will do it in this PR itself. Sounds good?

gmarull replied:
It’s ok, just raised as it’s always better to use existing facilities when available.

avisconti replied:
> It’s ok, just raised as it’s always better to use existing facilities when available.

Yes, correct. THanks for pointing this out. The stmemsc common APIs are now quite simpler.

gmarull commented on drivers/sensor/lis2ds12/lis2ds12.c:
```diff
@@ -267,32 +240,121 @@ static int lis2ds12_init(const struct device *dev)
+			   (void *)&lis2ds12_config_##inst.stmemsc_cfg,	\
+		},							\
+		.stmemsc_cfg.i2c = {					\
+			.bus = DEVICE_DT_GET(DT_INST_BUS(inst)),	\
+			.i2c_slv_addr = DT_INST_REG_ADDR(inst),		\
```
`I2C_DT_SPEC_INST_GET`

gmarull commented on drivers/sensor/lis2ds12/lis2ds12_trigger.c:
```diff
@@ -28,8 +24,8 @@ static void lis2ds12_gpio_callback(const struct device *dev,
 
-	gpio_pin_interrupt_configure(data->gpio, cfg->irq_pin,
-				     GPIO_INT_DISABLE);
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
```
this function should be checked for errors

avisconti replied:
But the callback returns void in this case. Do you want me to raise a LOG_ERR here?

gmarull replied:
Maybe an assert then?

avisconti replied:
sure

gmarull commented on drivers/sensor/lis2ds12/lis2ds12_trigger.c:
```diff
@@ -60,8 +55,8 @@ static void lis2ds12_handle_int(const struct device *dev)
 
-	gpio_pin_interrupt_configure(data->gpio, cfg->irq_pin,
-				     GPIO_INT_EDGE_TO_ACTIVE);
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
```
ditto

avisconti replied:
same here. LOG_ERR?

gmarull commented on drivers/sensor/lis2ds12/lis2ds12_trigger.c:
```diff
@@ -151,8 +160,8 @@ int lis2ds12_trigger_init(const struct device *dev)
 
-	gpio_pin_interrupt_configure(data->gpio, cfg->irq_pin,
-				     GPIO_INT_EDGE_TO_ACTIVE);
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
```
ditto

avisconti replied:
ack

gmarull commented on drivers/sensor/lis2ds12/lis2ds12_trigger.c:
```diff
@@ -162,14 +171,19 @@ int lis2ds12_trigger_set(const struct device *dev,
+		return -ENOTSUP;
+	}
+
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
```
ditto

avisconti replied:
ack

gmarull commented on drivers/sensor/lis2ds12/lis2ds12_trigger.c:
```diff
@@ -183,8 +197,8 @@ int lis2ds12_trigger_set(const struct device *dev,
 	lis2ds12_init_interrupt(dev);
-	gpio_pin_interrupt_configure(data->gpio, cfg->irq_pin,
-				     GPIO_INT_EDGE_TO_ACTIVE);
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
```
ditto

avisconti replied:
ack

gmarull reviewed (changes requested):
Changes lgtm, some minor comments that I think should be addressed

avisconti committed:
drivers/sensor: lis2ds12: Move odr Kconfig property into dts

Move odr options from Kconfigs to Device Tree. Moreover add
in DT a power-mode option to select among 4 possible values
(PD, LP, HR, HF). The power mode cannot be currently set from
sensor APIs.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented on drivers/sensor/lis2ds12/lis2ds12.c:
```diff
@@ -41,98 +41,65 @@ static struct lis2ds12_config lis2ds12_config = {
-		if (range == lis2ds12_accel_fs_map[i]) {
-			return i;
-		}
+	val =  LIS2DS12_HR_ODR_TO_REG(odr);
```
ack

avisconti commented on drivers/sensor/lis2ds12/lis2ds12.c:
```diff
@@ -41,98 +41,65 @@ static struct lis2ds12_config lis2ds12_config = {
+	val =  LIS2DS12_HR_ODR_TO_REG(odr);
+	if (val > LIS2DS12_XL_ODR_800Hz_HR) {
+		LOG_ERR("ODR too high");
+		return -ENOTSUP;
```
ack

avisconti commented on drivers/sensor/lis2ds12/lis2ds12.c:
```diff
@@ -41,98 +41,65 @@ static struct lis2ds12_config lis2ds12_config = {
-		return -EIO;
+	switch (range) {
+	default:
+	case 2:
```
ack

avisconti committed:
drivers/sensor: lis2ds12: Add multi-instance support

This commit aligns lis2ds12 sensor driver to latest multi
instance sensor driver model.

In particular it makes use of the stmemsc common routines
and move ctx handler inside struct config, so that the
bus_init routines can be totally avoided.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
drivers/sensor: stmemsc: make use of spi_dt_spec and i2c_dt_spec

Make use of the new DT facilities that introduced two new
bus structures, spi_dt_spec and i2c_dt_spec, as well as the
macros, SPI_DT_SPEC_INST_GET and I2C_DT_SPEC_INST_GET, to
retrieve info from DT.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

jefffhaynes commented:
@MaureenHelm @mbolivar-nordic I could really use this merge.  Any chance of getting some reviews?

avisconti commented:
> @MaureenHelm @mbolivar-nordic I could really use this merge. Any chance of getting some reviews?

@jefffhaynes I think this PR was blocked by v2.7.0 release.

MaureenHelm reviewed (commented):
LGTM.

@avisconti will you please rebase? I'd like to see a fresh CI run on this PR before merging.

avisconti commented:
Still running. Mmmmh...
